[BUILD] fix build of otlp_grpc_exporter_test with modern grpc

dbarker opened the pull request:
Fixes #3306

## Changes
- makes grpc link publicly to the grpc client 
- remove otlp_grpc_exporter_test's link to gRPC::grpc++
- removes extra protobuf links to tests 
- adds curl link to the zipkin test

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

owent reviewed (approved):
LGTM, thanks.
This issue was introduced by PR #3219. Could you please add some comments describing how to reproduce it?  
Otherwise, other contributors might reintroduce the same issue in the future, and we would have difficulty identifying it.

dbarker commented:
> LGTM, thanks. This issue was introduced by PR #3219. Could you please add some comments describing how to reproduce it? Otherwise, other contributors might reintroduce the same issue in the future, and we would have difficulty identifying it.

Thanks! Here are the steps to reproduce the build error in a linux container. 

1. build and run a dev container setting the latest grpc version with compatible protobuf and abseil versions. 
````
cd <path-to-opentelemetry-cpp>
docker build --build-arg GRPC_VERSION="v1.71.0" --build-arg PROTOBUF_VERSION="29.0" --build-arg ABSEIL_CPP_VERSION="20240722.1" --build-arg USER_UID="$(id -u)" --build-arg USER_GID="$(id -g)"  -t opentelemetry-cpp-dev:latest -f ./.devcontainer/Dockerfile.dev .
docker run -it -v "$PWD:/workspaces/opentelemetry-cpp" opentelemetry-cpp-dev:latest bash
````
2. build the otlp exporter and run tests
```
cmake -S . -B ~/build -DWITH_OTLP_GRPC=ON
cmake --build ~/build 
ctest --test-dir ~/build
```

I'll log an issue to upgrade to the latest release of vcpkg and use newer versions of grpc in ci. The link error solved by this PR should exist in grpc v1.67.0+.

dbarker commented:
The [build failure](https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/13996263060/job/39192105339?pr=3307#step:4:5135) is in an unrelated test of the otlp file exporter that [passed](https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/13984803212/job/39156548025#step:4:5069) in the previous run on the PR.

dbarker commented:
@owent  after more testing I found that the grpc link to the client should be public since the client is a public interface and exposes grpc headers. Making the link public also allows the tests to get the transitive dependency (solves a link error when building shared libraries on later versions of grpc). While I was in that file I've also removed extra protobuf links to tests since opentelemetry_proto exposes it publicly. I've tested this on old versions of grpc (1.49.2) and the latest (1.71.0). The zipkin exporter test needed curl and I've added it the branch.

owent commented on exporters/otlp/CMakeLists.txt:
```diff
@@ -38,18 +38,11 @@ if(WITH_OTLP_GRPC)
-      opentelemetry_exporter_otlp_grpc_client BEFORE
-      PUBLIC "$<BUILD_INTERFACE:${GRPC_INCLUDE_DIRECTORY}>")
-  endif()
+           opentelemetry_proto_grpc gRPC::grpc++
```
gRPC client definitions are not exposed to users, so it's fine to keep the gRPC dependency private. Alternatively, we could use `PUBLIC $<BUILD_INTERFACE:gRPC::grpc++> PRIVATE $<INSTALL_INTERFACE:gRPC::grpc++>` to expose the include directories internally. Symbol conflicts don't only cause issues during compilation or linking but will result in runtime crashes.

More details can be found at #1603 , #1891 , #1998

owent replied:
Do you think we can move `OtlpGrpcClientReferenceGuard` out of `otlp_grpc_client.h`, so we don't need to include `otlp_grpc_client.h` in headers of `OtlpGrpcClientFactory`?

dbarker replied:
I think we'd need to remove all grpc from `otlp_grpc_client.h` using pmpl and some generic status codes. Alternatively we could give the client a base class and use it in the constructors of the exporters and return from the factory - then deprecate the constructors and create method with the OtlpGrpcClient shard ptr.

dbarker replied:
actually #3321 is a good solution as long as the client methods are accessed outside the scope of the exporters. good call.

dbarker commented:
After more testing I found that building against grpc 1.71.0 is working on the current main branch. I'm not able to reproduce the link error originally reported [example here](https://github.com/dbarker/opentelemetry-cpp/actions/runs/13982683037/job/39150838265#step:5:1782).

I suspect the root issue was an ill-configured grpc install in my docker container and github runner. I had a typo on the arguments to the setup-grpc.sh script and `abseil` should be `abseil-cpp` https://github.com/open-telemetry/opentelemetry-cpp/blob/2c9b68fbaeef95bae9d3b5df20e6cf29f62c34d8/.devcontainer/Dockerfile.dev#L33

I'm going to close this PR and open separate PRs for: 
1. Making the grpc headers public in the installed otlp grpc client target (so grpc will not need to be linked explicit when creating a shared client) 
2. Fix the missing curl link to the zipkin exporter test
3. Fix the grpc install in the devcontainer docker file.

dbarker commented:
Closing.
